compiler/noirc_frontend/src/hir_def/types/unification.rs

```diff
@@ -0,0 +1,893 @@
+            }
+        }
+        false
+    }
+
+    /// Attempt to coerce `&mut T` to `&T`, returning true if this is possible.
+    pub fn try_reference_coercion(&self, target: &Type) -> bool {
+        let this = self.follow_bindings();
+        let target = target.follow_bindings();
+
+        if let (Type::Reference(this_elem, true), Type::Reference(target_elem, false)) =
+            (&this, &target)
+        {
+            // Still have to ensure the element types match.
+            // Don't need to issue an error here if not, it will be done in unify_with_coercions
+            let mut bindings = TypeBindings::default();
+            if this_elem.try_unify(target_elem, &mut bindings).is_ok() {
+                Self::apply_type_bindings(bindings);
+                return true;
+            }
+        }
+        false
+    }
+}
+
+/// Wraps a given `expression` in `expression.as_slice()`
+fn convert_array_expression_to_slice(
+    expression: ExprId,
+    array_type: Type,
+    target_type: Type,
+    as_slice_method: crate::node_interner::FuncId,
+    interner: &mut NodeInterner,
+) {
+    let as_slice_id = interner.function_definition_id(as_slice_method);
+    let location = interner.expr_location(&expression);
+    let as_slice = HirExpression::Ident(HirIdent::non_trait_method(as_slice_id, location), None);
+    let func = interner.push_expr(as_slice);
+
+    // Copy the expression and give it a new ExprId. The old one
+    // will be mutated in place into a Call expression.
+    let argument = interner.expression(&expression);
+    let argument = interner.push_expr(argument);
+    interner.push_expr_type(argument, array_type.clone());
+    interner.push_expr_location(argument, location);
+
+    let arguments = vec![argument];
+    let is_macro_call = false;
+    let call = HirExpression::Call(HirCallExpression { func, arguments, location, is_macro_call });
+    interner.replace_expr(&expression, call);
+
+    interner.push_expr_location(func, location);
+    interner.push_expr_type(expression, target_type.clone());
+
+    let func_type =
+        Type::Function(vec![array_type], Box::new(target_type), Box::new(Type::Unit), false);
+    interner.push_expr_type(func, func_type);
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::{BinaryTypeOperator, Kind, Type, TypeBindings, TypeVariable, TypeVariableId};
+
+    struct Types {
+        next_type_variable_id: usize,
+    }
+
+    impl Types {
+        fn new() -> Self {
+            Self { next_type_variable_id: 0 }
+        }
+
+        fn type_variable(&mut self) -> (Type, TypeVariableId) {
+            self.type_variable_with_kind(Kind::Any)
+        }
+
+        fn type_variable_with_kind(&mut self, kind: Kind) -> (Type, TypeVariableId) {
+            let id = TypeVariableId(self.next_type_variable_id);
+            self.next_type_variable_id += 1;
+            (Type::TypeVariable(TypeVariable::unbound(id, kind)), id)
+        }
+    }
+
+    fn num(value: u128) -> Type {
```

Comment:
```suggestion
    fn constant(value: u128) -> Type {
```
`num` makes me think `i32`, `u8`, `Field`, etc

Reply:
Done!